README
NAME

    Test::Spelling::Comment - check for spelling errors in code comments

VERSION

    Version 0.005

SYNOPSIS

        use Test::Spelling::Comment;
        Test::Spelling::Comment->new->add_stopwords(<DATA>)->all_files_ok;

DESCRIPTION

    Test::Spelling::Comment lets you check the spelling of your code
    comments, and report its results in standard Test::More fashion. This
    module uses Comment::Spell::Check to do the checking, which requires a
    spellcheck program such as spell, aspell, ispell, or hunspell.

    This test is an author test and should not run on end-user
    installations. Recommendation is to put it into your xt instead of your
    t directory.

USAGE

 new( [ ARGS ] )

    Returns a new Test::Spelling::Comment instance. new takes an optional
    hash with its arguments.

        Test::Spelling::Comment->new(
            skip      => pattern,
            stopwords => Pod::Wordlist,
        );

    The following arguments are supported:

  skip (optional)

    The skip argument is either a string or an array ref of strings or
    regex patterns. Every pattern is substituted for the empty string on
    every line of the input file. This happens before passing the file over
    to Comment::Spell::Check for spell checking.

    Use this option to remove parts of the file that would otherwise
    require you to add multiple stopwords. An example would be lines like
    these:

        # vim: ts=4 sts=4 sw=4 et: syntax=perl

  stopwords (optional)

    The stopwords argument must be a Pod::Wordlist instance, or something
    compatible. You can use that argument to configure Pod::Wordlist to
    your liking.

 file_ok( FILENAME )

    file_ok will ok the test and return something true if no spelling error
    is found in the code comments. Otherwise it fails the test and returns
    something false.

 all_files_ok

    Calls the all_files method of Test::XTFiles to get all the files to be
    tested. All files will be checked by calling file_ok.

    It calls done_testing or skip_all so you can't have already called
    plan.

    all_files_ok returns something true if all files test ok and false
    otherwise.

    Please see XT::Files for how to configure the files to be checked.

    WARNING: The API was changed with 0.005. Arguments to all_files_ok are
    now silently discarded and the method is now configured with XT::Files.

 add_stopwords( @entries )

    Adds the words passed in @entries as stopwords. These words are not
    passed to the spell checker and are therefore accepted as correct.

    The add_stopwords method always returns $self and can therefore be used
    to chain methods together.

    This method can be called multiple times.

    This method only adds the words as passed in @entries. Unlike
    learn_stopwords from Pod::Wordlist it does not add the words plural
    too.

EXAMPLES

 Example 1 Default usage

    Check the spelling in all files returned by XT::Files.

        use 5.006;
        use strict;
        use warnings;

        use Test::Spelling::Comment 0.002;

        if ( exists $ENV{AUTOMATED_TESTING} ) {
            print "1..0 # SKIP these tests during AUTOMATED_TESTING\n";
            exit 0;
        }

        Test::Spelling::Comment->new->add_stopwords(<DATA>)->all_files_ok;

        __DATA__
        your
        stopwords
        go
        here

 Example 2 Check non-default directories or files

    Use the same test file as in Example 1 and create a .xtfilesrc config
    file in the root directory of your distribution.

        [Dirs]
        module = lib
        module = tools
        module = corpus/hello

        [Files]
        module = corpus/my.pm

 Example 3 Call file_ok directly

        use 5.006;
        use strict;
        use warnings;

        use Test::More 0.88;
        use Test::Spelling::Comment;

        if ( exists $ENV{AUTOMATED_TESTING} ) {
            print "1..0 # SKIP these tests during AUTOMATED_TESTING\n";
            exit 0;
        }

        my $comment = Test::Spelling::Comment->new;
        $comment->file_ok('corpus/hello.pl');
        $comment->file_ok('tools/update.pl');

        done_testing();

 Example 4 Skip vim line

    Check the spelling in all files in the bin, script and lib directory
    and remove the vim comment.

        use 5.006;
        use strict;
        use warnings;

        use Test::Spelling::Comment 0.003;

        if ( exists $ENV{AUTOMATED_TESTING} ) {
            print "1..0 # SKIP these tests during AUTOMATED_TESTING\n";
            exit 0;
        }

        Test::Spelling::Comment->new(
            skip => '^# vim: .*'
        )->add_stopwords(<DATA>)->all_files_ok();

        __DATA__
        your
        stopwords
        go
        here

SEE ALSO

    Comment::Spell::Check, Comment::Spell, Test::More, XT::Files

SUPPORT

 Bugs / Feature Requests

    Please report any bugs or feature requests through the issue tracker at
    https://github.com/skirmess/Test-Spelling-Comment/issues. You will be
    notified automatically of any progress on your issue.

 Source Code

    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    https://github.com/skirmess/Test-Spelling-Comment

      git clone https://github.com/skirmess/Test-Spelling-Comment.git

AUTHOR

    Sven Kirmess <sven.kirmess@kzone.ch>



Changes
Revision history for Test-Spelling-Comment

{{$NEXT}}

  - Use redefine() instead of mock() from Test::MockModule
  - Switch license to the ISC license
  - Moved test files depending on additional modules to xt and skip them
    during installation testing
  - Removed test dependency on Test::Fatal

0.005  2019-08-03 23:39:53 UTC

  - API change: all_files_ok() no longer takes files or directories as
    arguments. The files to be tested is now obtained from XT::Files.

0.004  2018-06-25 18:13:37 UTC

  - Removed minimal dependency on Test::TempDir::Tiny 0.017
  - Use g flag for the substitution with the skip patterns

0.003  2018-06-24 19:35:33 UTC

  - added argument skip to new

0.002  2018-06-22 22:39:27 UTC

  - add_stopwords returns $self instead of undef to allow chaining
  - fixed abstract

0.001  2018-06-21 20:05:26 UTC

  - Initial release


t/lib/Local/Pod/Wordlist.pm
# vim: ts=4 sts=4 sw=4 et: syntax=perl
#
# Copyright (c) 2018-2022 Sven Kirmess
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

use 5.006;
use strict;
use warnings;

package Local::Pod::Wordlist;

our $VERSION = '0.001';

sub wordlist {
    return;
}

1;


t/lib/Local/Test/TempDir.pm
# vim: ts=4 sts=4 sw=4 et: syntax=perl
#
# Copyright (c) 2018-2022 Sven Kirmess
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

use 5.006;
use strict;
use warnings;

package Local::Test::TempDir;

our $VERSION = '0.001';

use Carp                  qw(croak);
use Cwd                   qw(getcwd);
use File::Path 2.07       qw(remove_tree);
use File::Spec::Functions qw(catdir);

# Support Exporter < 5.57
require Exporter;
our @ISA       = qw(Exporter);    ## no critic (ClassHierarchies::ProhibitExplicitISA)
our @EXPORT_OK = qw(tempdir);

{
    my $temp_dir_base;

    sub _init {
        return if defined $temp_dir_base;

        my $root_dir;
        if ( exists $ENV{LOCAL_TEST_TEMPDIR_BASEDIR} ) {
            $root_dir = $ENV{LOCAL_TEST_TEMPDIR_BASEDIR};
            croak "env variables LOCAL_TEST_TEMPDIR_BASEDIR doesn't point to a valid directory: $root_dir" if !-d $root_dir;
        }
        else {
            $root_dir = getcwd();
            croak "Cannot get cwd: $!" if !defined $root_dir;
        }

        $temp_dir_base = catdir( $root_dir, 'tmp' );
        if ( !-d $temp_dir_base ) {
            mkdir $temp_dir_base or croak "Cannot create directory $temp_dir_base $!";
        }

        ( my $dirname = $0 ) =~ tr{:\\/.}{_};
        $temp_dir_base = catdir( $temp_dir_base, $dirname );
        if ( !-e $temp_dir_base ) {
            mkdir $temp_dir_base or croak "Cannot create directory $temp_dir_base $!";
        }
        elsif ( -l $temp_dir_base || !-d _ ) {
            croak "Not a directory $temp_dir_base";
        }
        else {
            remove_tree( $temp_dir_base, { safe => 0, keep_root => 1 } );
        }

        return;
    }

    my %counter;

    sub tempdir {
        my $label = defined( $_[0] ) ? $_[0] : 'default';
        $label =~ tr{a-zA-Z0-9_-}{_}cs;

        if ( exists $counter{$label} ) {
            $counter{$label}++;
        }
        else {
            $counter{$label} = '0';
        }

        $label = "${label}_$counter{$label}";

        _init();

        my $tempdir = catdir( $temp_dir_base, $label );
        mkdir $tempdir or croak "Cannot create directory: $!";

        return $tempdir;
    }
}

1;


lib/Test/Spelling/Comment.pm
# vim: ts=4 sts=4 sw=4 et: syntax=perl
#
# Copyright (c) 2018-2022 Sven Kirmess
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

use 5.006;
use strict;
use warnings;

package Test::Spelling::Comment;

our $VERSION = '0.005';

use Moo;

use Carp                  ();
use Comment::Spell::Check ();
use Pod::Wordlist         ();
use Scalar::Util          ();
use Test::Builder         ();
use Test::XTFiles         ();

has _skip => (
    is       => 'ro',
    init_arg => 'skip',
);

has _stopwords => (
    is        => 'ro',
    isa       => sub { Carp::croak q{stopwords must have method 'wordlist'} if !Scalar::Util::blessed( $_[0] ) || !$_[0]->can('wordlist'); },
    init_arg  => 'stopwords',
    lazy      => 1,
    default   => sub { Pod::Wordlist->new },
    predicate => 1,
);

my $TEST = Test::Builder->new();

# - Do not use subtests because subtests cannot be tested with
#   Test::Builder:Tester.
# - Do not use a plan because a method that sets a plan cannot be tested
#   with Test::Builder:Tester.
# - Do not call done_testing in a method that should be tested by
#   Test::Builder::Tester because TBT cannot test them.

sub add_stopwords {
    my $self = shift;

    my $wordlist = $self->_stopwords->wordlist;

  STOPWORD:
    for (@_) {

        # explicit copy
        my $stopword = $_;
        $stopword =~ s{ ^ \s* }{}xsm;
        $stopword =~ s{ \s+ $ }{}xsm;
        next STOPWORD if $stopword eq q{};

        $wordlist->{$stopword} = 1;
    }

    return $self;
}

sub all_files_ok {
    my ($self) = @_;

    my @files = Test::XTFiles->new->all_files();
    if ( !@files ) {
        $TEST->skip_all("No files found\n");
        return 1;
    }

    my $rc = 1;
    for my $file (@files) {
        if ( !$self->file_ok($file) ) {
            $rc = 0;
        }
    }

    $TEST->done_testing;

    return 1 if $rc;
    return;
}

sub file_ok {
    my ( $self, $file ) = @_;

    Carp::croak 'usage: file_ok(FILE)' if @_ != 2 || !defined $file;

    if ( !-f $file ) {
        $TEST->ok( 0, $file );
        $TEST->diag("\n");
        $TEST->diag("File $file does not exist or is not a file");

        return;
    }

    my $fh;
    if ( !open $fh, '<', $file ) {
        $TEST->ok( 0, $file );
        $TEST->diag("\n");
        $TEST->diag("Cannot read file '$file': $!");

        return;
    }

    my @lines = <$fh>;
    chomp @lines;

    if ( !close $fh ) {
        $TEST->ok( 0, $file );
        $TEST->diag("\n");
        $TEST->diag("Cannot read file '$file': $!");

        return;
    }

    my $skips_ref = $self->_skip;
    if ( defined $skips_ref ) {
        if (   ( !defined Scalar::Util::reftype($skips_ref) )
            || ( Scalar::Util::reftype($skips_ref) ne Scalar::Util::reftype( [] ) ) )
        {
            $skips_ref = [$skips_ref];
        }

        for my $line (@lines) {
            for my $skip ( @{$skips_ref} ) {
                ## no critic (RegularExpressions::RequireDotMatchAnything)
                ## no critic (RegularExpressions::RequireExtendedFormatting)
                ## no critic (RegularExpressions::RequireLineBoundaryMatching)
                $line =~ s{$skip}{}g;
                ## use critic
            }
        }
    }

    my $speller = Comment::Spell::Check->new( $self->_has_stopwords ? ( stopwords => $self->_stopwords ) : () );
    my $buf;
    $speller->set_output_string($buf);
    my $result;
    if ( !eval { $result = $speller->parse_from_string( join "\n", @lines, q{} ); 1 } ) {
        my $error_msg = $@;
        $TEST->ok( 0, $file );
        $TEST->diag("\n$error_msg\n\n");

        return;
    }

    if ( @{ $result->{fails} } == 0 ) {
        $TEST->ok( 1, $file );

        return 1;
    }

    $TEST->ok( 0, $file );
    $TEST->diag("\n$buf\n\n");

    return;
}

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::Spelling::Comment - check for spelling errors in code comments

=head1 VERSION

Version 0.005

=head1 SYNOPSIS

    use Test::Spelling::Comment;
    Test::Spelling::Comment->new->add_stopwords(<DATA>)->all_files_ok;

=head1 DESCRIPTION

C<Test::Spelling::Comment> lets you check the spelling of your code
comments, and report its results in standard L<Test::More|Test::More>
fashion. This module uses L<Comment::Spell::Check|Comment::Spell::Check> to
do the checking, which requires a spellcheck program such as C<spell>,
C<aspell>, C<ispell>, or C<hunspell>.

This test is an author test and should not run on end-user installations.
Recommendation is to put it into your F<xt> instead of your F<t> directory.

=head1 USAGE

=head2 new( [ ARGS ] )

Returns a new C<Test::Spelling::Comment> instance. C<new> takes an optional
hash with its arguments.

    Test::Spelling::Comment->new(
        skip      => pattern,
        stopwords => Pod::Wordlist,
    );

The following arguments are supported:

=head3 skip (optional)

The C<skip> argument is either a string or an array ref of strings or regex
patterns. Every pattern is substituted for the empty string on every line of
the input file. This happens before passing the file over to
L<Comment::Spell::Check|Comment::Spell::Check> for spell checking.

Use this option to remove parts of the file that would otherwise require you
to add multiple C<stopwords>. An example would be lines like these:

    # vim: ts=4 sts=4 sw=4 et: syntax=perl

=head3 stopwords (optional)

The C<stopwords> argument must be a L<Pod::Wordlist|Pod::Wordlist> instance,
or something compatible. You can use that argument to configure
L<Pod::Wordlist|Pod::Wordlist> to your liking.

=head2 file_ok( FILENAME )

C<file_ok> will ok the test and return something I<true> if no spelling
error is found in the code comments. Otherwise it fails the test and returns
something I<false>.

=head2 all_files_ok

Calls the C<all_files> method of L<Test::XTFiles> to get all the files to
be tested. All files will be checked by calling C<file_ok>.

It calls C<done_testing> or C<skip_all> so you can't have already called
C<plan>.

C<all_files_ok> returns something I<true> if all files test ok and I<false>
otherwise.

Please see L<XT::Files> for how to configure the files to be checked.

WARNING: The API was changed with 0.005. Arguments to C<all_files_ok>
are now silently discarded and the method is now configured with
L<XT::Files>.

=head2 add_stopwords( @entries )

Adds the words passed in C<@entries> as stopwords. These words are not
passed to the spell checker and are therefore accepted as correct.

The C<add_stopwords> method always returns C<$self> and can therefore be
used to chain methods together.

This method can be called multiple times.

This method only adds the words as passed in C<@entries>. Unlike
C<learn_stopwords> from L<Pod::Wordlist|Pod::Wordlist> it does not add the
words plural too.

=head1 EXAMPLES

=head2 Example 1 Default usage

Check the spelling in all files returned by L<XT::Files>.

    use 5.006;
    use strict;
    use warnings;

    use Test::Spelling::Comment 0.002;

    if ( exists $ENV{AUTOMATED_TESTING} ) {
        print "1..0 # SKIP these tests during AUTOMATED_TESTING\n";
        exit 0;
    }

    Test::Spelling::Comment->new->add_stopwords(<DATA>)->all_files_ok;

    __DATA__
    your
    stopwords
    go
    here

=head2 Example 2 Check non-default directories or files

Use the same test file as in Example 1 and create a F<.xtfilesrc> config
file in the root directory of your distribution.

    [Dirs]
    module = lib
    module = tools
    module = corpus/hello

    [Files]
    module = corpus/my.pm

=head2 Example 3 Call C<file_ok> directly

    use 5.006;
    use strict;
    use warnings;

    use Test::More 0.88;
    use Test::Spelling::Comment;

    if ( exists $ENV{AUTOMATED_TESTING} ) {
        print "1..0 # SKIP these tests during AUTOMATED_TESTING\n";
        exit 0;
    }

    my $comment = Test::Spelling::Comment->new;
    $comment->file_ok('corpus/hello.pl');
    $comment->file_ok('tools/update.pl');

    done_testing();

=head2 Example 4 Skip vim line

Check the spelling in all files in the F<bin>, F<script> and F<lib>
directory and remove the C<vim> comment.

    use 5.006;
    use strict;
    use warnings;

    use Test::Spelling::Comment 0.003;

    if ( exists $ENV{AUTOMATED_TESTING} ) {
        print "1..0 # SKIP these tests during AUTOMATED_TESTING\n";
        exit 0;
    }

    Test::Spelling::Comment->new(
        skip => '^# vim: .*'
    )->add_stopwords(<DATA>)->all_files_ok();

    __DATA__
    your
    stopwords
    go
    here

=head1 SEE ALSO

L<Comment::Spell::Check|Comment::Spell::Check>,
L<Comment::Spell|Comment::Spell>, L<Test::More|Test::More>,
L<XT::Files>

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/skirmess/Test-Spelling-Comment/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software. The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/skirmess/Test-Spelling-Comment>

  git clone https://github.com/skirmess/Test-Spelling-Comment.git

=head1 AUTHOR

Sven Kirmess <sven.kirmess@kzone.ch>

=cut


Makefile.PL
use 5.006;
use strict;
use warnings;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
    ABSTRACT           => 'check for spelling errors in code comments',
    AUTHOR             => 'Sven Kirmess <sven.kirmess@kzone.ch>',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    DISTNAME         => 'Test-Spelling-Comment',
    LICENSE          => 'open_source',
    MIN_PERL_VERSION => '5.006',
    NAME             => 'Test::Spelling::Comment',
    PREREQ_PM        => {
        Carp                    => 0,
        'Comment::Spell::Check' => 0,
        Moo                     => 0,
        'Pod::Wordlist'         => 0,
        'Scalar::Util'          => 0,
        'Test::Builder'         => 0,
        'Test::XTFiles'         => 0,
        strict                  => 0,
        warnings                => 0,
    },
    TEST_REQUIRES => {
        Cwd                     => 0,
        Exporter                => 0,
        'File::Basename'        => 0,
        'File::Path'            => '2.07',
        'File::Spec'            => 0,
        'File::Spec::Functions' => 0,
        'Test::Builder::Tester' => 0,
        'Test::More'            => '0.88',
        lib                     => 0,
        subs                    => 0,
    },
    VERSION => '0.005',
    test    => {
        TESTS => 't/*.t',
    },
);

my %FallbackPrereqs = (
    Carp                    => 0,
    'Comment::Spell::Check' => 0,
    Cwd                     => 0,
    Exporter                => 0,
    'File::Basename'        => 0,
    'File::Path'            => '2.07',
    'File::Spec'            => 0,
    'File::Spec::Functions' => 0,
    Moo                     => 0,
    'Pod::Wordlist'         => 0,
    'Scalar::Util'          => 0,
    'Test::Builder'         => 0,
    'Test::Builder::Tester' => 0,
    'Test::More'            => '0.88',
    'Test::XTFiles'         => 0,
    lib                     => 0,
    strict                  => 0,
    subs                    => 0,
    warnings                => 0,
);

if ( $ENV{AUTOMATED_TESTING} || $ENV{EXTENDED_TESTING} ) {
    $WriteMakefileArgs{test}{TESTS} .= ' xt/*.t';
    _add_extended_test_requirements();
}

if ( !eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
    delete $WriteMakefileArgs{TEST_REQUIRES};
    delete $WriteMakefileArgs{BUILD_REQUIRES};
    $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

if ( !eval { ExtUtils::MakeMaker->VERSION(6.52) } ) {
    delete $WriteMakefileArgs{CONFIGURE_REQUIRES};
}

WriteMakefile(%WriteMakefileArgs);

sub test_requires {
    my ( $module, $version_or_range ) = @_;
    $WriteMakefileArgs{TEST_REQUIRES}{$module} = $FallbackPrereqs{$module} = $version_or_range;
    return;
}

sub _add_extended_test_requirements {
    test_requires( 'Test::MockModule', '0.14' );
    test_requires( 'constant',         0 );
    return;
}

# vim: ts=4 sts=4 sw=4 et: syntax=perl
